src/Components/Components/src/DynamicComponent.cs

```diff
@@ -0,0 +1,104 @@
+
+        /// <summary>
+        /// Constructs an instance of <see cref="DynamicComponent"/>.
+        /// </summary>
+        public DynamicComponent()
+        {
+            _cachedRenderFragment = Render;
+        }
+
+        /// <summary>
+        /// Gets or sets the type of the component to be rendered. The supplied type must
+        /// implement <see cref="IComponent"/>.
+        /// </summary>
+        [Parameter]
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)]
+        public Type Type { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets a dictionary of parameters to be passed to the component.
+        /// </summary>
+        // Note that this deliberately does *not* use CaptureUnmatchedValues. Reasons:
+        // [1] The primary scenario for DynamicComponent is where the call site doesn't
+        //     know which child component it's rendering, so it typically won't know what
+        //     set of parameters to pass either, hence the developer most likely wants to
+        //     pass a dictionary rather than having a fixed set of parameter names in markup.
+        // [2] If we did have CaptureUnmatchedValues here, then it would become a breaking
+        //     change to ever add more parameters to DynamicComponent itself in the future,
+        //     because they would shadow any coincidentally same-named ones on the target
+        //     component. This could lead to application bugs.
+        [Parameter]
+        public IDictionary<string, object>? Parameters { get; set; }
+
+        /// <inheritdoc />
+        public void Attach(RenderHandle renderHandle)
+        {
+            _renderHandle = renderHandle;
+        }
+
+        /// <inheritdoc />
+        public Task SetParametersAsync(ParameterView parameters)
+        {
+            // This manual parameter assignment logic will be marginally faster than calling
+            // ComponentProperties.SetProperties.
+            foreach (var entry in parameters)
+            {
+                if (entry.Name.Equals(nameof (Type), StringComparison.OrdinalIgnoreCase))
+                {
+                    Type = (Type)entry.Value;
+                }
+                else if (entry.Name.Equals(nameof(Parameters), StringComparison.OrdinalIgnoreCase))
+                {
+                    Parameters = (IDictionary<string, object>)entry.Value;
+                }
+                else
+                {
+                    throw new InvalidOperationException(
+                        $"{nameof(DynamicComponent)} does not accept a parameter with the name '{entry.Name}'. To pass parameters to the dynamically-rendered component, use the '{nameof(Parameters)}' parameter.");
+                }
+            }
+
+            _renderHandle.Render(_cachedRenderFragment);
+            return Task.CompletedTask;
+        }
+
+        void Render(RenderTreeBuilder builder)
+        {
+            if (Type is null)
```

Comment:
Any reason not to do this as part of `SetParametersAsync`?

Reply:
In virtually every imaginable scenario it won't make any difference, neither to behavior or to the number of operations executed. So mostly I think it doesn't matter.

If we're willing to stretch plausibility a bit, we could theoretically imagine a scenario where, in between `SetParametersAsync` and the `Render` delegate being invoked, the developer changed `Type` to be `null`. In this scenario, you would prefer the behavior to be what we have here, i.e., that we complain about the null only at the point where we're trying to use the null.

Overall though I don't think it makes any real difference here.

Reply:
Update: I needed to submit another commit to restart CI which seems to have got stuck, so took the opportunity to move this check as suggested. On consideration I think your expectation that this happens in `SetParametersAsync` would most likely line up with other people's expectations, so best to avoid surprise.